### PR TITLE
tests/feature_resolution: add build system unit tests

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Detecting bugs in feature resolution previously resulted in a lot of manual testing and more than once an issue slipped through. Dedicated unit tests should fix both.

### Testing procedure

~~Hopefully, it is indeed as trivial as I had hoped for and Murdock will just run the test.~~ The issue described in https://github.com/RIOT-OS/RIOT/pull/15967#pullrequestreview-586857807 should be detected by this test.

Update: The test is now part of `make static-test`.

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/15957, https://github.com/RIOT-OS/RIOT/pull/15855

### CODE REVIEW

**Comment by maribu** [comment]
Static test is failing as expected. May I squash?

File: dist/tools/ci/static_tests.sh [hunk 1]
```diff
@@ -122,6 +122,7 @@ run ./dist/tools/coccinelle/check.sh
 run ./dist/tools/flake8/check.sh
 run ./dist/tools/headerguards/check.sh
 run ./dist/tools/buildsystem_sanity_check/check.sh
+run ./tests/feature_resolution/check.sh
```
Comment (fjmolinas) [inline dist/tools/ci/static_tests.sh:125]: Wouldn't this fit better in `dist/tools/...`?

File: tests/feature_resolution/Makefile.test [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  $(info EXPECTED_FEATURES_USED_BLACKLISTED = $(EXPECTED_FEATURES_USED_BLACKLISTED))
+  $(info FEATURES_USED_BLACKLISTED = $(FEATURES_USED_BLACKLISTED))
+  $(error FEATURES_USED_BLACKLISTED incorrectly resolved)
+endif
+
+ifneq ($(EXPECTED_FEATURES_CONFLICTING),$(FEATURES_CONFLICTING))
+  $(info EXPECTED_FEATURES_CONFLICTING = $(EXPECTED_FEATURES_CONFLICTING))
+  $(info FEATURES_CONFLICTING = $(FEATURES_CONFLICTING))
+  $(error FEATURES_CONFLICTING incorrectly resolved)
+endif
```
Comment (fjmolinas) [inline tests/feature_resolution/Makefile.test:25]:
```suggestion
FEATURES_VARS := \
  FEATURES_CONFLICTING \
  FEATURES_USED \
  FEATURES_USED_BLACKLISTED \
  FEATURES_MISSING \
  #

define _expected =
  ifneq ($$(EXPECTED_$1),$$($1))
    $$(info EXPECTED_$1 = $$(EXPECTED_$1))
    $$(info $1= $$($1))
    $$(error $1 incorrectly resolved)
  endif
endef

$(foreach var,$(FEATURES_VARS),$(eval $(call _expected,$(var))))

```

File: tests/feature_resolution/check.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+cd "$(dirname "$0")"
```
Comment (fjmolinas) [inline tests/feature_resolution/check.sh:2]:
```suggestion
cd "$(dirname "$0")" || return 1
```

**Review by fjmolinas — COMMENTED** [review]
Nice and needed addition. Maybe these files fit better under `makefiles/tests`?

**Comment by fjmolinas** [comment]
Can we add this test case as well

```
# inputs
FEATURES_OPTIONAL := a
FEATURES_REQUIRED :=
FEATURES_REQUIRED_ANY := d|b|a
FEATURES_PROVIDED := a b c
FEATURES_BLACKLIST :=
FEATURES_CONFLICT :=

# expected results
EXPECTED_FEATURES_USED := a
EXPECTED_FEATURES_MISSING :=
EXPECTED_FEATURES_USED_BLACKLISTED :=
EXPECTED_FEATURES_CONFLICTING :=

include Makefile.test
```

**Comment by maribu** [comment]
Another bug uncovered: blacklisted features end still up being used if they are optional.

**Comment by maribu** [comment]
@fjmolinas: Ping?

**Comment by fjmolinas** [comment]
Lets spin up the ci

**Comment by fjmolinas** [comment]
I'm running the dependency check, although I expect differences since this PR caught some bugs, Probably will have some bugs in `gnrc_minimal` since without this PR both `libc` get selected.

**Comment by fjmolinas** [comment]
But

> I'm running the dependency check, although I expect differences since this PR caught some bugs, Probably will have some bugs in `gnrc_minimal` since without this PR both `libc` get selected.

So I just want to take a look but won't be a blocking point since this PR provides tests.

**Review by fjmolinas — APPROVED** [review]
Everything seems ok, lets go, hopefully, there are no more hidden bugs lingering.

**Comment by maribu** [comment]
Thanks!

**Pull request merged**